dhis2/dhis2-core
# feat: email verification

# Summary
Adds a verified email field to the User object, to facilitate for email verification.
The users email will be verified when the user has received an email with a verification link and clicked it. When the link is clicked, the user's verified email field will be set to the same as the email field.
The email verified field is unique in contrast to the original email field, hence many can have the same email field value, but only one can have the verified email.

Technical non-obvious choices made:

-  This is the first iteration of this feature to facilitate specifically for 2FA emails. 
- Having two email fields, instead of only one and a boolean saying verified/not verified.
   - In order support both unique emails and non-unique emails for backward compatibility, only one, the `verifiedEmail` is unique.
- Another column `emailverificationtoken` is used for storing the verification token.
- An account is not "locked" during/after verification email is sent/requested. You can send as many verification emails as you want, until the latest is clicked/verified and the state is changed to verified. You can change email and re-verify, by first changing the non-unique email, then resend the verification and verify the new one. There is no automation to this procedure in order to be backward compatible and verification has to be triggered/requested manually trough the API for now.
-  In this first iteration of this feature, we only intend to use this as a prerequisite for enabling 2FA with email, i.e. you can't enable 2FA with email, unless your email is verified first. 
- No restrictions is added to the functionality of other features based on verified vs non-verified emails.
- You can try to verify any email you want, you can change your email after verification and try to verify the new one.
- When changing your non-unique email, you are automatically not having a verified email, until you verify the new one.
- System wide enforcement of verified emails is not a part of this task, but can be added in a later iteration.
- We are only forcing users to verify their email if and ONLY when they want to enable 2FA with email.
- There is no automatic/background cleanup of expired tokens, they will just be overwritten next time you try to send a new verification email, and set to NULL on successful verification.



## Automatic tests
AccountControllerTest

## Manual test
There is no UI for this feature yet.
To test manually. Do the following API calls on an instance with a valid SMTP server configured:
1. Make sure your test user has en email set, that you control. Use this user when calling the API.
2. Call  POST `/api/users/USER_ID/sendEmailVerification`
3. Open your email and click the link or call: GET `/api/users/USER_ID/verifyEmail?token=YOUR_TOKEN_RECEIVED_IN_THE_LINK`
4. Observe that the user now has the verifiedEmail field set: GET `/api/users/USER_ID`

### JIRA:
[DHIS2-16853](https://dhis2.atlassian.net/browse/DHIS2-16853)

[DHIS2-16853]: https://dhis2.atlassian.net/browse/DHIS2-16853?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
